MdePkg/Library/BaseCacheMaintenanceLib/RiscVCache.c

```diff
@@ -11,51 +12,116 @@
 #include <Library/DebugLib.h>
 
 /**
-  RISC-V invalidate instruction cache.
-
+  Use runtime discovery mechanism in future when avalable
+  through https://lists.riscv.org/g/tech-privileged/topic/83853282
 **/
-VOID
-EFIAPI
-RiscVInvalidateInstCacheAsm (
-  VOID
-  );
+#define RV64_CACHE_BLOCK_SIZE   64
+
+#define __stringify_1(x...)	#x
+#define __stringify(x...)	__stringify_1(x)
+
+#define CMO_OP(_op, _start)\
+asm volatile("cbo." __stringify(_op) " (%0)" :: "r" (Start))

```

Comment:
inline assembly is discouraged in edk2. Could you convert it into assembly functions?

Reply:
Fixed. Also it seems like we will require to do refactoring in BaseLib.h when your first set of refactored patches are checked in.

Reply:
What kind of refactoring do we need? Do you mean splitting the header file or the library?

Reply:
Please check my comments in [/RiscVCache.c](https://github.com/dhaval-rivos/edk2/pull/2/commits/c7c7cb1c92a809331661975102a4f2a713733915#diff-b3b4a147e01dbef8fc0f87cbb130f1577a8f9158dbd648578d4b074508496739)

Reply:
But why is it dependent on refactoring work? BaseLib already has RiscV64/FlushCache.S, right? Why can't these new assembly routines be added there?

Reply:
If you compare x86cache.c and Riscvcache.c: x86cache uses AsmWbinvd declared in Baselib.h and implemented in MdePkg/Baselib. Idea was probably that Baselib will have all common functions to manage cache. But RV has slightly different functionality so we can not use same baselib.h declarations and instead I am linking them directly back into MdePkh/Baselib (without using baselib.h)

Reply:
During refactoring if we can consider AsmFlushCacheLine and other relevant cache management APIs considering RV, then it will be easy to maintain the same links.